Fix race conditions in HIP ParallelScan

seyonglee opened the pull request:
This PR adds missing synchronizations to the HIP ParallelScan implementation to fix race conditions.
This fixes the issue #8607

seyonglee committed:
Add missing synchronizations to the HIP ParallelScan implementation to
fix race conditions.

Signed-off-by: Seyong Lee <lees2@ornl.gov>

masterleinad commented:
Can you explain for every synchronization barrier added why they are necessary?

masterleinad commented:
Since the HIP implementation for the most part is a copy of the Cuda implementation, I would expect that we need the same barriers for both backends.

JBludau commented:
Do we know why this did not show up on other AMD gpus?

tcclevenger reviewed (commented):
With adding this many syncs, do we expect a big performance difference?

tcclevenger commented on core/src/HIP/Kokkos_HIP_ParallelScan_Range.hpp:
```diff
@@ -169,7 +170,7 @@ class ParallelScanHIPBase {
       }
 
       // Make sure the write is seen by all threads
-      __threadfence_block();
```
What is the reason to change this to a `__syncthreads()`?

seyonglee replied:
`__threadfence_block()` flushes memory to enforce memory ordering, but it does not provide an implicit synchronization. Therefore, it is still possible that before thread A finishes the memory fence operation, thread B executes the next statement that reads the same shared memory that thread A tries to flush.

JBludau replied:
and __syncthreads implies __threadfence, see https://rocm.docs.amd.com/projects/HIP/en/latest/how-to/hip_cpp_language_extensions.html#synchronization-functions

tcclevenger commented on core/src/HIP/Kokkos_HIP_ParallelScan_Range.hpp:
```diff
@@ -202,6 +203,7 @@ class ParallelScanHIPBase {
                 reinterpret_cast<pointer_type>(shared_prefix)),
             true);
       }
+      __syncthreads();
```
I reverted every change in the PR, except this `__syncthread()`, and it fixed the error in the reproducer from https://github.com/kokkos/kokkos/issues/8607. At least in that specific case, this seems to be the only change needed.

seyonglee replied:
Race condition can occur if two threads writes and read the same shared memory; therefore, I conservatively added `__syncthreads()` whenever shared memory is written by a thread but read by different thread. However, actual race condition will occur only when the two write/read threads are in different scheduling units (different warps).
I will check which `__syncthreads()` can be safely removed. (BTW, the race condition occurs more frequently if we test with very large inputs- bigger than the example in #8607)

seyonglee replied:
Deleted unnecessary `__syncthreads()`.

masterleinad replied:
This looks strange. We are executing the functor for the same memory address when calling the functor and when writing the memory out in the next line. No other thread should access that memory in the meantime.

masterleinad replied:
@masterleinad Can you coimment here?

JBludau replied:
I guess you ment to write @tcclevenger

masterleinad replied:
Either way, it's not clear to me why `__syncthreads` makes sense here.

tcclevenger replied:
It could be anything from line 174 to here that would require a sync. I just removed this one last, so it triggered the error, so maybe there is something else earlier?

seyonglee replied:
With all the recent changes and clarifications, I also don't know why this synchronization is needed; I will keep digging into this.

crtrott replied:
agreed that this guy doesn't really make sense.

seyonglee replied:
Changing `__syncthreads()` to `__threadfenc_block()` here seems to work too, but I don't know why.

seyonglee committed:
Remove unnecessary __syncthreads().
Signed-off-by: Seyong Lee <lees2@ornl.gov>

Signed-off-by: Seyong Lee <lees2@ornl.gov>

seyonglee commented on core/src/HIP/Kokkos_HIP_ReduceScan.hpp:
```diff
@@ -382,7 +382,7 @@ __device__ bool hip_single_inter_block_reduce_scan_impl(
       global[i] = shared[i];
     }
-    __threadfence();
+    __syncthreads();
```
`__threadfence()` does not provide necessary synchronization; therefore, changed to `__syncthreads()`.

masterleinad replied:
Makes sense.

Rombur replied:
Are we sure about this change? AMD gave us this code when they added the MI300 backend in https://github.com/kokkos/kokkos/pull/6358

seyonglee replied:
You are right; we don't need to change `__threadfence()` to `__syncthreads()` here since the next `__syncthreads_or()` provides necessary synchronization using `atomicInc` (the last block implicitly waits for all previous block finishes).

seyonglee commented:
> Do we know why this did not show up on other AMD gpus?

The race condition may occur only if a thread block consists more than a single warp and GPU provides aggressive warp/thread scheduling.

masterleinad commented on core/src/HIP/Kokkos_HIP_ParallelScan_Range.hpp:
```diff
@@ -169,7 +169,7 @@ class ParallelScanHIPBase {
 
       // Make sure the write is seen by all threads
-      __threadfence_block();
+      __syncthreads();
```
Cuda has `__syncwarp(MASK);` here so it's reasonable that `__threadfence_block` isn't strong enough.

Rombur replied:
You need `__syncwarp` in CUDA because of the Independent Thread Scheduling. For AMD, all the threads are executed in lock steps, so there is no need to wait for all the threads in the warp. I am not convince this change is necessary.

seyonglee replied:
> You need `__syncwarp` in CUDA because of the Independent Thread Scheduling. For AMD, all the threads are executed in lock steps, so there is no need to wait for all the threads in the warp. I am not convince this change is necessary.

If none of AMD GPUs supports independent thread scheduling, yes, we don't need to change this.
(It seems that MI300A does not support the independent thread scheduling - the test succeeds without this change.)

Rombur replied:
None of AMD GPUs supports ITS. That's why the HIP backend looks more like the pre-Volta CUDA code than the current implementation.

seyonglee replied:
Then, I will undo this change; thanks!

seyonglee committed:
Change __syncthreads() back to __threadfence() in line 385 of
Kokkos_HIP_ReduceScan.hpp
Signed-off-by: Seyong Lee <lees2@ornl.gov>

Signed-off-by: Seyong Lee <lees2@ornl.gov>

seyonglee committed:
Change `__syncthreads()` back to `__threadfence_block()`.
AMD GPUs do not support the independent thread scheduling, and thus
`__threadfence_block()` will be enough.
Signed-off-by: Seyong Lee <lees2@ornl.gov>

Signed-off-by: Seyong Lee <lees2@ornl.gov>

seyonglee committed:
Changed  __syncthreads() to __threadfenc_block()

Signed-off-by: Seyong Lee <lees2@ornl.gov>

masterleinad commented:
```diff
diff --git a/core/src/HIP/Kokkos_HIP_ParallelScan_Range.hpp b/core/src/HIP/Kokkos_HIP_ParallelScan_Range.hpp
index ce9b35b0d..93f23dd4a 100644
--- a/core/src/HIP/Kokkos_HIP_ParallelScan_Range.hpp
+++ b/core/src/HIP/Kokkos_HIP_ParallelScan_Range.hpp
@@ -156,9 +156,6 @@ class ParallelScanHIPBase {
          iwork_base < range.end(); iwork_base += blockDim.y) {
       const typename Policy::member_type iwork = iwork_base + threadIdx.y;
 
-      __syncthreads();  // Don't overwrite previous iteration values until they
-                        // are used
-
       final_reducer.init(
           reinterpret_cast<pointer_type>(shared_prefix + word_count.value));
 
@@ -205,6 +202,8 @@ class ParallelScanHIPBase {
       if (iwork + 1 == m_policy.end() && m_policy.end() == range.end() &&
           m_result_ptr_device_accessible)
         *m_result_ptr = *reinterpret_cast<pointer_type>(shared_prefix);
+      // Don't overwrite previous iteration values until they are used
+      __syncthreads();
     }
   }
 ```
Also fixes it. For some reason, having a memory barrier at the end of the loop is important.

masterleinad commented:
Also note that the unit test doesn't use the result anyway.

seyonglee committed:
Update as suggested by @masterleinad

Signed-off-by: Seyong Lee <lees2@ornl.gov>

seyonglee commented:
> ```diff
> diff --git a/core/src/HIP/Kokkos_HIP_ParallelScan_Range.hpp b/core/src/HIP/Kokkos_HIP_ParallelScan_Range.hpp
> index ce9b35b0d..93f23dd4a 100644
> --- a/core/src/HIP/Kokkos_HIP_ParallelScan_Range.hpp
> +++ b/core/src/HIP/Kokkos_HIP_ParallelScan_Range.hpp
> @@ -156,9 +156,6 @@ class ParallelScanHIPBase {
>           iwork_base < range.end(); iwork_base += blockDim.y) {
>        const typename Policy::member_type iwork = iwork_base + threadIdx.y;
>  
> -      __syncthreads();  // Don't overwrite previous iteration values until they
> -                        // are used
> -
>        final_reducer.init(
>            reinterpret_cast<pointer_type>(shared_prefix + word_count.value));
>  
> @@ -205,6 +202,8 @@ class ParallelScanHIPBase {
>        if (iwork + 1 == m_policy.end() && m_policy.end() == range.end() &&
>            m_result_ptr_device_accessible)
>          *m_result_ptr = *reinterpret_cast<pointer_type>(shared_prefix);
> +      // Don't overwrite previous iteration values until they are used
> +      __syncthreads();
>      }
>    }
> ```
> 
> Also fixes it. For some reason, having a memory barrier at the end of the loop is important.

Updated as suggested.

masterleinad reviewed (commented):
We should also add a test case.

seyonglee committed:
Add a new unit test to test parallel_scan() with a custom scalar variable.

Signed-off-by: Seyong Lee <lees2@ornl.gov>

Adrien-Tab commented on core/unit_test/TestParallelScanRangePolicyCustomScalar.hpp:
```diff
@@ -0,0 +1,348 @@
+};  // struct TestParallelScanRangePolicyCustomScalar
+
+TEST(TEST_CATEGORY, parallel_scan_range_policy_customscalar) {
+  using scalar_type = ArrayValueType<size_t, 2>;
```
This could also be tested with `ArrayValueType<double, 2>`

seyonglee replied:
Added.

Adrien-Tab commented on core/unit_test/TestParallelScanRangePolicyCustomScalar.hpp:
```diff
@@ -0,0 +1,348 @@
+
+      for (size_t i = 0; i < work_size; ++i) {
+        // Check prefix sum
+        ASSERT_EQ(i * (i - 1) / 2, prefix_h(i).v[0]);
```
Shouldn't `v[1]` be tested as well?

seyonglee replied:
Added.

masterleinad commented on core/unit_test/TestParallelScanRangePolicyCustomScalar.hpp:
```diff
@@ -0,0 +1,348 @@
+  }
+}
+}  // namespace
+
```
```suggestion
}  // namespace
```

seyonglee committed:
Apply suggestion from @masterleinad

Co-authored-by: Daniel Arndt <arndtd@ornl.gov>
Signed-off-by: Seyong Lee <lees2@ornl.gov>

seyonglee committed:
Add v[1] checking code as suggested by @AdRi1t

Signed-off-by: Seyong Lee <lees2@ornl.gov>

seyonglee committed:
Add testing with ArrayValueType<double, 2> as suggested by @AdRi1t

Signed-off-by: Seyong Lee <lees2@ornl.gov>

seyonglee commented:
> We should also add a test case.

Added.

masterleinad reviewed (approved):
Approving so SNL-CI can run.

Adrien-Tab commented:
The CI fail for this test on 32-bit (x86) + OpenMP 
linux / CI (x86, ubuntu:latest, g++, RelWithDebInfo, OPENMP, 20, -m32, -DKokkos_ENABLE_HWLOC=OFF)
```
/__w/kokkos/kokkos/core/unit_test/TestParallelScanRangePolicyCustomScalar.hpp:118: Failure
Expected equality of these values:
  i * (i + 1) / 2
    Which is: 32768
  postfix_h(i).v[0]
    Which is: 2147516416
```

masterleinad committed:
Disable tests failing for 32-bit build

dalg24 commented on core/unit_test/TestParallelScanRangePolicyCustomScalar.hpp:
```diff
@@ -0,0 +1,394 @@
+namespace {
+
+template <typename T, int N>
+struct ArrayValueType {
```
Is that guy copied from elsewhere?
It might be worth cleaning it up.
For one it conflicts with #8689

seyonglee replied:
Updated the constructors.

dalg24 commented on core/unit_test/TestParallelScanRangePolicyCustomScalar.hpp:
```diff
@@ -0,0 +1,394 @@
+
+        // Check postfix sum
+        ASSERT_EQ(i * (i + 1) / 2, postfix_h(i).v[0]);
+        ASSERT_EQ(i * (i + 1) / 2, postfix_h(i).v[1]);
```
This is quite brittle an assumes that `ValueType` is `ArrayValueType<T, 2>`

seyonglee replied:
Updated to use a loop to check each element.

dalg24 commented on core/src/HIP/Kokkos_HIP_ParallelScan_Range.hpp:
```diff
@@ -156,9 +156,6 @@ class ParallelScanHIPBase {
 
-      __syncthreads();  // Don't overwrite previous iteration values until they
-                        // are used
-
```
Do we have the same issue with Cuda?

seyonglee replied:
Yes, synchronization between iterations are necessary for both Cuda and HIP.

crtrott reviewed (changes requested):
This fix doesn't make any sense to me. By moving the fence, you removed a required fence after the init and before the first work iteration, and you added a fence at the end of the last work iteration, with no code coming after (its the last thing in the kernel). I don't see anything happening after the "final()" call after all.

crtrott commented:
> Copilot encountered an error and was unable to review this pull request. You can try again by re-requesting a review.

Hu, interesting.

seyonglee commented:
> This fix doesn't make any sense to me. By moving the fence, you removed a required fence after the init and before the first work iteration, and you added a fence at the end of the last work iteration, with no code coming after (its the last thing in the kernel). I don't see anything happening after the "final()" call after all.

In the `iwork_base` loop in line 155, we need `__syncthreads()` before moving to next iteration, but the first iteration may not need `__syncthreads()` for the previous `shared_accum` initialization if the target GPU does not support Independent Thread Scheduling (ITS), which is the case for AMD GPUs. (`shared_accum` is initialized by threads 0 .. (word_count.value - 1), which will be in the same warp (scheduling unit).)
The reason for moving `__syncthreads()` from the beginning of the `iwork_base` loop to the end of the loop is because experiments require extra memory fence at the end of the loop (even though I still don't understand the reason for that). Moving `__synchthread()` gives the memory fence effect at the end of the loop while not increasing the total number of `__syncthreads()` in the loop. 

CF: If `word_count.value`> warp-size, we will need `__synchthreads()` before the first iteration too.

seyonglee committed:
Update ArrayValueType constructors.

Signed-off-by: Seyong Lee <lees2@ornl.gov>

crtrott commented:
> experiments require extra memory fence at the end of the loop (even though I still don't understand the reason for that)

That is my point. We need a fence at the beginning, and it doesn't make sense to me that we need one at the very end.

crtrott commented:
So if there is really something wrong somewhere with the synchronization at the end of a kernel, leave the syncthreads at the beginning and add one after the loop.

seyonglee committed:
Add __syncthreads() to make initial writes to shared_accum visible to
threads in a thread block.

Signed-off-by: Seyong Lee <lees2@ornl.gov>

seyonglee commented:
> So if there is really something wrong somewhere with the synchronization at the end of a kernel, leave the syncthreads at the beginning and add one after the loop.

Moving syncthreads to the beginning of the loop and adding another synchthreads after the loop still causes errors.
Instead, kept the syncthreads at the end of the loop and added another syncthreads before the loop to make the initial writes to `shared_accum` visible.

seyonglee committed:
Update result-compare code as suggested.

Signed-off-by: Seyong Lee <lees2@ornl.gov>

science-enthusiast commented on core/unit_test/TestParallelScanRangePolicyCustomScalar.hpp:
```diff
@@ -0,0 +1,391 @@
+  ArrayValueType& operator=(const T& b) {
+    for (int i = 0; i < N; ++i) this->v[i] = b;
+    return *this;
+  }
```
Shouldn't the overload for `operator=` based on `const T&` be enough? I think the only drawback is that it will lead to copying for both lvalues and rvalues. In case we want to retain the other overload for move semantics, we will have to forward `b` in the body of that function. Also, it should be `T&&`.

seyonglee replied:
Agree; deleted the unnecessary constructor.

science-enthusiast commented on core/unit_test/TestParallelScanRangePolicyCustomScalar.hpp:
```diff
@@ -0,0 +1,391 @@
+  using ViewType  = Kokkos::View<ValueType*, execution_space>;
+
+  ViewType prefix_results;
+  ViewType postfix_results;
```
Please don't mind me saying this. Let us avoid refering to exclusive scan as postfix scan. We can call the variables related to inclusive scan and exclusive scan as `in_scan_*` and `ex_scan_*`, for example.

seyonglee replied:
Renamed as suggested.

science-enthusiast replied:
Thank you. I looked more closely at the code now and I see that prefix was exclusive scan and postfix was inclusive scan.

seyonglee committed:
Delete unnecessary constructor, as suggested.

Signed-off-by: Seyong Lee <lees2@ornl.gov>

seyonglee committed:
Rename prefix scan to exclusive scan and postfix scan to inclusive scan, as suggested.

Signed-off-by: Seyong Lee <lees2@ornl.gov>

JBludau commented:
@seyonglee @crtrott what is the status on this? This might be something we want sorted out soon

seyonglee commented:
> @seyonglee @crtrott what is the status on this? This might be something we want sorted out soon

The current status is that we added one extra synchronization (and moved one existing synchronization) for correct execution of the new test, but we still don't know why the extra synchronization is needed.

JBludau commented:
> > @seyonglee @crtrott what is the status on this? This might be something we want sorted out soon
> 
> The current status is that we added one extra synchronization (and moved one existing synchronization) for correct execution of the new test, but we still don't know why the extra synchronization is needed.

Thx. Nevertheless, I think we should move forward with this but keep investigating since users might get wrong results without noticing

seyonglee committed:
Add comments to Kokkos_HIP_ParallelScan_Range.hpp
Delete TestParallelScanRangePolicyCustomScalar.hpp
Add a simpler unit test: TestCustomScalarParallelScan.hpp

Signed-off-by: Seyong Lee <lees2@ornl.gov>

seyonglee commented:
Add comments explaining which parts are modified in the ParallelScan implementation.
Changed the unit test with a simpler one, which reproduces errors more frequently.

masterleinad commented on core/unit_test/TestCustomScalarParallelScan.hpp:
```diff
@@ -0,0 +1,92 @@
+    accum += rowptr_h(i);
+  }
+
+  //[FIXME] below causes a random test failure.
```
```suggestion
```

masterleinad commented:
```
/builds/kokkos/kokkos/core/unit_test/TestCustomScalarParallelScan.hpp(74): error: The enclosing parent function ("TestBody") for an extended __host__ __device__ lambda cannot have private or protected access within its class
        [=] __attribute__((host)) __attribute__((device))(const int& lr, update_type& update, const bool& final) {
                                                         ^
```
You need to delegate the test using `Kokkos::parallel_scan` with `KOKKOS_LAMBDA` to separate function to make `nvcc` happy.

seyonglee committed:
Update core/unit_test/TestCustomScalarParallelScan.hpp

Co-authored-by: Daniel Arndt <arndtd@ornl.gov>
Signed-off-by: Seyong Lee <lees2@ornl.gov>

seyonglee committed:
Delegate the test to a separate function, as suggested, to avoid the extended lambda issue on the CUDA backend

Signed-off-by: Seyong Lee <lees2@ornl.gov>

seyonglee commented:
> ```
> /builds/kokkos/kokkos/core/unit_test/TestCustomScalarParallelScan.hpp(74): error: The enclosing parent function ("TestBody") for an extended __host__ __device__ lambda cannot have private or protected access within its class
>         [=] __attribute__((host)) __attribute__((device))(const int& lr, update_type& update, const bool& final) {
>                                                          ^
> ```
> 
> You need to delegate the test using `Kokkos::parallel_scan` with `KOKKOS_LAMBDA` to separate function to make `nvcc` happy.

Thank you for catching this; updated.

crtrott committed:
Update comments indicating compiler issue on MI300A

Signed-off-by: Christian Trott <crtrott@sandia.gov>
